pkg/util/ansible_execution.go

```diff
@@ -98,51 +108,87 @@ func AnsibleExecution(
 			ansibleEE.Spec.Playbook = service.Spec.Playbook
 		}
 
-		ansibleEEMounts := storage.VolMounts{}
-		sshKeyVolume := corev1.Volume{
-			Name: "ssh-key",
-			VolumeSource: corev1.VolumeSource{
-				Secret: &corev1.SecretVolumeSource{
-					SecretName: sshKeySecret,
-					Items: []corev1.KeyToPath{
-						{
-							Key:  "ssh-privatekey",
-							Path: "ssh_key",
+		// If we have a service that ought to be deployed everywhere
+		// substitute the existing play target with 'all'
+		// Check if we have ExtraVars before accessing it
+		if ansibleEE.Spec.ExtraVars == nil {
+			ansibleEE.Spec.ExtraVars = make(map[string]json.RawMessage)
+		}
+		if service.Spec.DeployOnAllNodeSets != nil && *service.Spec.DeployOnAllNodeSets {
+			ansibleEE.Spec.ExtraVars["edpm_override_hosts"] = json.RawMessage([]byte("\"all\""))
+			util.LogForObject(helper, fmt.Sprintf("for service %s, substituting existing ansible play host with 'all'.", service.Name), ansibleEE)
```

Comment:
so if we have two nodesets and a global service then what will happen?
* A) run the ansible playbook of the service once with hosts=all
* B) run the ansible playbook of the service twice both with hosts=all

what will happen if both of the two nodesets have the global service added?
what will happen if only one of the two nodesets have the global service added?

Reply:
Good point. Services don't have any concept of dependencies, yet, so B is our answer.

That isn't necessarily catastrophic, because our services use ansible to make changes. If the roles are written properly, utilizing idempotent operations, or at least operations that don't cause damage if run twice, we should be fine.

Furthermore, just like with standard services, we rely on user to make the right call when setting up nodesets. Global services add a new aspect to this. But it's not that different from the current situation, when we expect users not to include one service twice in a nodeset, or to run services which are mutually incompatible.

That being said. I'll add documentation about this.

Reply:
I guess the concern here is about conflicts if there were 2 global executions simultaneously. It would be 2 different ansible processes trying to configure known_hosts at the same time, which could cause an issue. We could document this as a limitation, and add a check to make sure it doesn't occur, as we are already iterating over all the services, we'd need to check that each global service is only specified once. If we wanted to go beyond a check, we could just make sure the service is only execute one time, by making sure it only runs once by modifying the list of services we sent to the deploy function.

Reply:
We can check that the global service is specified only once. But I don't think we should resolve wrong specification for the user by modifying the list. There should either be an error, preventing deployment completely, or a warning.

Reply:
Yeah the main concern is about simultaneously executing both instance of the global service against the same node. Do we deploy nodesets in a deployment sequentially or in parallel? If the latter, then we need to do some logic in the operator for mutual exclusiveness of the two run of the same service at the same time, if the former then doc is enough.

Reply:
Btw, what will be our doc strategy here?
* ask the user to add the global service to each nodeset in the deployment
* ask the user to add the global service to one of the nodesets and ensure that if that nodeset is removed later then the global service is added to another nodeset in the deployment

Reply:
Have global service in one deployment. It's going to get executed against all.

Reply:
but the service is in nodeset not in deployment.